src/tokens/ERC1155B.sol

```diff
@@ -0,0 +1,266 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity >=0.8.0;
+
+import {ERC1155TokenReceiver} from "./ERC1155.sol";
+
+/// @notice Minimalist and gas efficient ERC1155 implementation optimized for single supply ids.
+/// @author Solmate (https://github.com/Rari-Capital/solmate/blob/main/src/tokens/ERC1155B.sol)
+abstract contract ERC1155B {
+    /*//////////////////////////////////////////////////////////////
+                                 EVENTS
+    //////////////////////////////////////////////////////////////*/
+
+    event TransferSingle(
+        address indexed operator,
+        address indexed from,
+        address indexed to,
+        uint256 id,
+        uint256 amount
+    );
+
+    event TransferBatch(
+        address indexed operator,
+        address indexed from,
+        address indexed to,
+        uint256[] ids,
+        uint256[] amounts
+    );
+
+    event ApprovalForAll(address indexed owner, address indexed operator, bool approved);
+
+    event URI(string value, uint256 indexed id);
+
+    /*//////////////////////////////////////////////////////////////
+                             ERC1155 STORAGE
+    //////////////////////////////////////////////////////////////*/
+
+    mapping(address => mapping(address => bool)) public isApprovedForAll;
+
+    /*//////////////////////////////////////////////////////////////
+                            ERC1155B STORAGE
+    //////////////////////////////////////////////////////////////*/
+
+    mapping(uint256 => address) public ownerOf;
+
+    function balanceOf(address owner, uint256 id) public view returns (uint256 bal) {
+        address idOwner = ownerOf[id];
+
+        assembly {
+            // We avoid branching by using assembly to take
+            // the bool output of eq() and use it as a uint.
+            bal := eq(idOwner, owner)
+        }
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                             METADATA LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    function uri(uint256 id) public view virtual returns (string memory);
+
+    /*//////////////////////////////////////////////////////////////
+                              ERC1155 LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    function setApprovalForAll(address operator, bool approved) public virtual {
+        isApprovedForAll[msg.sender][operator] = approved;
+
+        emit ApprovalForAll(msg.sender, operator, approved);
+    }
+
+    function safeTransferFrom(
+        address from,
+        address to,
+        uint256 id,
+        uint256 amount,
+        bytes memory data
+    ) public virtual {
+        require(msg.sender == from || isApprovedForAll[from][msg.sender], "NOT_AUTHORIZED");
```

Comment:
This comment applies to all `require` operations: why not use custom errors? I mean it decreases deployment costs for sure as well as runtime costs if the revert condition is met. Also, I think including the address of the emitter in all custom errors would make sense. This helps end users identify which contract was reverted with a failed transaction, which is especially useful for complex transactions involving multiple contracts:

```solidity
error SomeError(address emitter, ..., ..., ...);

emit SomeError(address(this), ..., ..., ...);
```

But probably you're already opinionated about custom errors 🤪